### PR TITLE
Refactor ChangePassword

### PR DESCRIPTION
**Pull request by madmax330**
### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/211242

### Tests

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

- Log into an account on NewDot
- Go to Profile > Security > Change Password
- Enter a wrong password and a new password
- Make sure you see the error displayed and that the api response contains the onyxData

```
{"onyxData":[{"onyxMethod":"merge","key":"account","value":{"error":"Your password changes could not be saved because the old password entered is incorrect."}}],"jsonCode":200,"requestID":"mV4I24"}
```

- Enter the correct password and a new password and everything should work.

- [x] Verify that no errors appear in the JS console

### CODE REVIEW

**Review by techievivek — COMMENTED** [review]
Looks good so far. Is there anything else remaining to be updated with this other than updating the PR checklist and adding tests and videos/screenshots?

**Comment by madmax330** [comment]
PR is ready for review, should be reviewed and tested alongside: https://github.com/Expensify/Web-Expensify/pull/33885

pulling in @iwiznia since you said you had time for refactor reviews 😄

File: src/libs/actions/User.js [hunk 1]
```diff
@@ -46,25 +47,34 @@ Onyx.connect({
...
-            }
-
-            const success = lodashGet(response, 'message', 'Password changed successfully.');
-            Onyx.merge(ONYXKEYS.ACCOUNT, {success});
-        })
-        .finally(() => {
-            Onyx.merge(ONYXKEYS.ACCOUNT, {loading: false});
-        });
+function changePassword(oldPassword, password) {
+    const optimisticData = [
```
Comment (iwiznia) [inline src/libs/actions/User.js:52]: NAB: I don't think this variables are needed or make things much clearer

File: src/pages/settings/PasswordPage.js [hunk 2]
```diff
@@ -133,7 +133,7 @@ class PasswordPage extends Component {
         if (!this.validate()) {
             return;
         }
-        User.changePasswordAndNavigate(this.state.currentPassword, this.state.newPassword);
```
Comment (iwiznia) [inline src/pages/settings/PasswordPage.js:136]: So, this never really navigated?

**Review by iwiznia — DISMISSED** [review]

File: src/pages/settings/PasswordPage.js (hunk 2 above)
Comment (madmax330) [reply src/pages/settings/PasswordPage.js:136]: Apparently not 🤷🏽

Comment (mountiny) [reply src/pages/settings/PasswordPage.js:136]: Yeah, seems like the original function never had any navigation in it 😄

**Review by mountiny — COMMENTED** [review]
Came over out of curiosity, LGTM. It would great to leave a video of the testing steps succeeding with the Web-E branch.

**Review by techievivek — COMMENTED** [review]
Changes seem to work fine. Can we add a video testing this new refactor? 

Also While testing this, I observed something interesting(not sure if it is a bug) but once I changed the password (and got a confirmation) I clicked on the back arrow and tried changing the password again and it shows `Session expired error` and this is nowhere visible on the UI just in the response data and spinner is in running state. But I didn't have to log in again I refreshed the page and everything worked fine.

Screenshot 2022-06-07 at 5 13 59 PM

**Comment by madmax330** [comment]
Refactored to use the new command. Here is the video:

https://user-images.githubusercontent.com/7016692/172562142-c38e644d-9a35-4508-b860-073845ce9b3d.mov


> Also While testing this, I observed something interesting(not sure if it is a bug) but once I changed the password (and got a confirmation) I clicked on the back arrow and tried changing the password again and it shows Session expired error and this is nowhere visible on the UI just in the response data and spinner is in running state. But I didn't have to log in again I refreshed the page and everything worked fine.

Hmmm, that's interesting, although I don't think it's related to this PR.

**Review by mountiny — COMMENTED** [review]
> Also While testing this, I observed something interesting(not sure if it is a bug) but once I changed the password (and got a confirmation) I clicked on the back arrow and tried changing the password again and it shows Session expired error and this is nowhere visible on the UI just in the response data and spinner is in running state. But I didn't have to log in again I refreshed the page and everything worked fine.

@techievivek would you be able to see if the same reproduction steps do this on the main too or it does not happen on main branch? To be sure it is related to this PR

**Comment by techievivek** [comment]
While testing something on NewDot observed something which I think this PR is missing. Once when we are done with changing the password when we try changing the password **again** it does give us `Session expired error` but then it makes a subsequent call to `Authenticate` command and then recalls the `ChangePassword` command which is not the same for this PR as it keeps calling `ChangePassword` which returns `Session expired error`.(All this is done on main branch)
Screenshot 2022-06-12 at 8 43 01 PM

**Comment by madmax330** [comment]
Hmm the only explanation I can think of is b/c we're using API.write() instead of the deprecated API.

**Comment by madmax330** [comment]
Either way, I'm waiting on the web PR to be complete then I'll test this again and try to figure it out.

**Comment by marcaaron** [comment]
One idea is to just force all requests in the write queue to have a `shouldRetry` true if they are called via `API.write()`

https://github.com/Expensify/App/blob/07fabcf6c2b033faedb5215e2c9ddcc195fe98fc/src/libs/Network/index.js#L38-L39

Looks like these params are no longer included, but the reauthentication middleware still looks at them.

**Comment by madmax330** [comment]
Alright, updated and retested, everything should be fine now

File: src/libs/API.js [hunk 3]
```diff
@@ -19,7 +19,14 @@ function write(command, apiCommandParameters = {}, onyxData = {}) {
     // Assemble all the request data we'll be storing in the queue
     const request = {
         command,
-        data,
+        data: {
+            ...data,
+
+            // We add this here because this was passed in the old api requests and some middlewares still expect this.
+            // This should be removed once we are no longer using deprecatedAPI
```
Comment (marcaaron) [inline src/libs/API.js:26]: Let's create a follow up issue to do this vs. leaving this "todo" in the code.

File: src/libs/API.js [hunk 4]
```diff
@@ -19,7 +19,14 @@ function write(command, apiCommandParameters = {}, onyxData = {}) {
     // Assemble all the request data we'll be storing in the queue
     const request = {
         command,
-        data,
+        data: {
+            ...data,
+
+            // We add this here because this was passed in the old api requests and some middlewares still expect this.
```
Comment (marcaaron) [inline src/libs/API.js:25]: NAB, this comment is not necessary. Whoever wants to remove this can just look to see if it's being used and why.

**Review by marcaaron — COMMENTED** [review]
When testing I am seeing multiple API calls and calls to `UpdatePassword` and a call to `Authenticate` (even though the authToken is up to date). I might need to update Auth or something though or maybe it is expected that a successful call to `UpdatePassword` returns this response? Unsure...

**Comment by marcaaron** [comment]
https://user-images.githubusercontent.com/32969087/174670367-41401540-ee36-4d38-b629-cf0bef9f2435.mp4

**Comment by madmax330** [comment]
Yeah @marcaaron what you posted is the expected behavior. 
The first call to `UpdatePassword` returns 407 (unauthenticated) so it authenticates, then calls `UpdatePassword` again. 
Same as prod.

**Comment by techievivek** [comment]
Yep, that is expected behaviour.

**Review by techievivek — APPROVED** [review]
Looks good to me now. Thanks

**Comment by marcaaron** [comment]
Ok I was confused but I think what happens is....

- Call `UpdatePassword` once (returns 200 but your authToken gets invalidated)
- Call `UpdatePassword` a second time and you will have to first reauthenticate and then call `UpdatePassword` again

**Review by marcaaron — APPROVED** [review]

**Comment by marcaaron** [comment]
Linked Web-E PR is on production so gonna merge this one 🚀

**Pull request merged**